### PR TITLE
Dial with block in clienturl

### LINKED ISSUE #755: TestConnectServer_RequestParallel is not stable
```
2021-03-05T07:01:40.5038110Z --- FAIL: TestConnectServer_RequestParallel (4.60s)
2021-03-05T07:01:40.5042250Z     server_test.go:304: 
2021-03-05T07:01:40.5042770Z         	Error Trace:	server_test.go:304
2021-03-05T07:01:40.5043240Z         	            				asm_amd64.s:1374
2021-03-05T07:01:40.5043740Z         	Error:      	Received unexpected error:
2021-03-05T07:01:40.5044300Z         	            	context deadline exceeded
2021-03-05T07:01:40.5044960Z         	Test:       	TestConnectServer_RequestParallel
2021-03-05T07:01:40.5045590Z     server_test.go:304: 
2021-03-05T07:01:40.5046050Z         	Error Trace:	server_test.go:304
2021-03-05T07:01:40.5046500Z         	            				asm_amd64.s:1374
2021-03-05T07:01:40.5047010Z         	Error:      	Received unexpected error:
2021-03-05T07:01:40.5047570Z         	            	context deadline exceeded
2021-03-05T07:01:40.5048680Z         	Test:       	TestConnectServer_RequestParallel
...
...
...
2021-03-05T07:01:42.0394020Z         	Error Trace:	server_test.go:304
2021-03-05T07:01:42.0394470Z         	            				asm_amd64.s:1374
2021-03-05T07:01:42.0394970Z         	Error:      	Received unexpected error:
2021-03-05T07:01:42.0395520Z         	            	context deadline exceeded
2021-03-05T07:01:42.0396300Z         	Test:       	TestConnectServer_RequestParallel
2021-03-05T07:01:42.0396890Z     server_test.go:329: 
2021-03-05T07:01:42.0397340Z         	Error Trace:	server_test.go:329
2021-03-05T07:01:42.0397790Z         	Error:      	Not equal: 
2021-03-05T07:01:42.0398210Z         	            	expected: 1000
2021-03-05T07:01:42.0398610Z         	            	actual  : 0
2021-03-05T07:01:42.0399170Z         	Test:       	TestConnectServer_RequestParallel
2021-03-05T07:01:42.0405920Z ##[error]    leaks.go:78: found unexpected goroutines:
2021-03-05T07:01:42.0419560Z         [Goroutine 15 in state semacquire, with sync.runtime_Semacquire on top of the stack:
2021-03-05T07:01:42.0420280Z         goroutine 15 [semacquire]:
2021-03-05T07:01:42.0420820Z         sync.runtime_Semacquire(0xc0000c5158)
2021-03-05T07:01:42.0421540Z         	/Users/runner/hostedtoolcache/go/1.15.8/x64/src/runtime/sema.go:56 +0x45
2021-03-05T07:01:42.0422170Z         sync.(*WaitGroup).Wait(0xc0000c5150)
2021-03-05T07:01:42.0422890Z         	/Users/runner/hostedtoolcache/go/1.15.8/x64/src/sync/waitgroup.go:130 +0xe5
2021-03-05T07:01:42.0427690Z         github.com/networkservicemesh/sdk/pkg/networkservice/common/connect_test.TestConnectServer_RequestParallel.func2(0x1919560, 0xc0003801b0, 0xc000364bd0, 0xc00031d080, 0xc0000c5140, 0xc0000c5150, 0x191fee0, 0xc00037e980, 0x26a)
2021-03-05T07:01:42.0429200Z         	/Users/runner/work/sdk/sdk/pkg/networkservice/common/connect/server_test.go:309 +0x2e5
2021-03-05T07:01:42.0430480Z         created by github.com/networkservicemesh/sdk/pkg/networkservice/common/connect_test.TestConnectServer_RequestParallel
2021-03-05T07:01:42.0431690Z         	/Users/runner/work/sdk/sdk/pkg/networkservice/common/connect/server_test.go:292 +0x796
2021-03-05T07:01:42.0432240Z         
2021-03-05T07:01:42.0432900Z          Goroutine 723 in state semacquire, with sync.runtime_Semacquire on top of the stack:
2021-03-05T07:01:42.0433590Z         goroutine 723 [semacquire]:
2021-03-05T07:01:42.0434130Z         sync.runtime_Semacquire(0xc0000c5158)
2021-03-05T07:01:42.0434850Z         	/Users/runner/hostedtoolcache/go/1.15.8/x64/src/runtime/sema.go:56 +0x45
2021-03-05T07:01:42.0435480Z         sync.(*WaitGroup).Wait(0xc0000c5150)
2021-03-05T07:01:42.0436180Z         	/Users/runner/hostedtoolcache/go/1.15.8/x64/src/sync/waitgroup.go:130 +0xe5
2021-03-05T07:01:42.0437760Z         github.com/networkservicemesh/sdk/pkg/networkservice/common/connect_test.TestConnectServer_RequestParallel.func2(0x1919560, 0xc0003801b0, 0xc000364bd0, 0xc00031d080, 0xc0000c5140, 0xc0000c5150, 0x191fee0, 0xc00037e980, 0x26d)
2021-03-05T07:01:42.0439310Z         	/Users/runner/work/sdk/sdk/pkg/networkservice/common/connect/server_test.go:309 +0x2e5
2021-03-05T07:01:42.0440530Z         created by github.com/networkservicemesh/sdk/pkg/networkservice/common/connect_test.TestConnectServer_RequestParallel
2021-03-05T07:01:42.0441710Z         	/Users/runner/work/sdk/sdk/pkg/networkservice/common/connect/server_test.go:292 +0x796
...
...
...
2021-03-05T07:01:50.2159540Z          Goroutine 1101 in state semacquire, with sync.runtime_Semacquire on top of the stack:
2021-03-05T07:01:50.2160190Z         goroutine 1101 [semacquire]:
2021-03-05T07:01:50.2160720Z         sync.runtime_Semacquire(0xc0000c5158)
2021-03-05T07:01:50.2161530Z         	/Users/runner/hostedtoolcache/go/1.15.8/x64/src/runtime/sema.go:56 +0x45
2021-03-05T07:01:50.2162200Z         sync.(*WaitGroup).Wait(0xc0000c5150)
2021-03-05T07:01:50.2162870Z         	/Users/runner/hostedtoolcache/go/1.15.8/x64/src/sync/waitgroup.go:130 +0xe5
2021-03-05T07:01:50.2164410Z         github.com/networkservicemesh/sdk/pkg/networkservice/common/connect_test.TestConnectServer_RequestParallel.func2(0x1919560, 0xc0003801b0, 0xc000364bd0, 0xc00031d080, 0xc0000c5140, 0xc0000c5150, 0x191fee0, 0xc00037e980, 0x3e7)
2021-03-05T07:01:50.2165820Z         	/Users/runner/work/sdk/sdk/pkg/networkservice/common/connect/server_test.go:309 +0x2e5
2021-03-05T07:01:50.2167040Z         created by github.com/networkservicemesh/sdk/pkg/networkservice/common/connect_test.TestConnectServer_RequestParallel
2021-03-05T07:01:50.2168220Z         	/Users/runner/work/sdk/sdk/pkg/networkservice/common/connect/server_test.go:292 +0x796
2021-03-05T07:01:50.2168760Z         ]
2021-03-05T07:01:50.2169090Z No logger in context
2021-03-05T07:01:50.2169480Z No logger in context
2021-03-05T07:01:50.2169860Z No logger in context
2021-03-05T07:01:50.2170240Z No logger in context
2021-03-05T07:01:50.2170560Z FAIL
2021-03-05T07:01:50.2171220Z FAIL	github.com/networkservicemesh/sdk/pkg/networkservice/common/connect	5.619s
```

### PR DESCRIPTION
**Pull request by Bolodya1997**
# Issue
`clienturl` client does dial with `dialTimeout` but without any `WithBlock`/`WithReturnConnectionError` option, so such timeout doesn't make any sense.
# Solution
Append `WithReturnConnectionError` to the given dial options.

Additionally this PR should provide more descriptive errors for #755 failures.

## Update
1. Extracts `dialTimeout` to `clienturl.NewClient`, `connect.NewServer` parameters to solve #755 issue.
2. `TestNSMGR_SelectsRestartingEndpoint` doesn't really selects **restarting** endpoint, fixed this.

Fixes #755.

### CODE REVIEW

File: pkg/networkservice/chains/nsmgr/server.go [hunk 1]
```diff
@@ -116,7 +116,7 @@ func NewServer(ctx context.Context, nsmRegistration *registryapi.NetworkServiceE
 		recvfd.NewServer(), // Receive any files passed
 		interpose.NewServer(&interposeRegistryServer),
 		filtermechanisms.NewServer(&urlsRegistryServer),
-		connect.NewServer(ctx,
+		connect.NewServer(ctx, 100*time.Millisecond,
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgr/server.go:119]: I feel this should be an optional parameter.

File: pkg/networkservice/chains/nsmgr/server_test.go [hunk 2]
```diff
@@ -131,14 +135,32 @@ func TestNSMGR_SelectsRestartingEndpoint(t *testing.T) {
...
 		NetworkServiceNames: []string{"ns-1"},
-	}
-
-	_, err := domain.Nodes[0].NewEndpoint(ctx, nseReg, sandbox.GenerateTestToken, &restartingEndpoint{startTime: time.Now().Add(time.Second * 2)})
+		Url:                 "tcp://" + netListener.Addr().String(),
+	})
 	require.NoError(t, err)
 
+	// 2. Postpone endpoint start
+	time.AfterFunc(2*time.Second, func() {
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgr/server_test.go:157]:
```suggestion
	time.AfterFunc(time.Second, func() {
```

File: pkg/networkservice/common/clienturl/client.go [hunk 3]
```diff
@@ -46,9 +48,10 @@ type clientURLClient struct {
 // NewClient - creates a Client that will using clienturl.ClientUrl(ctx) to extract a url, dial it to a cc, use that cc with the clientFactory to produce a new
 //             client to which it passes through any Request or Close calls
 // 	ctx	- is full lifecycle context, any started clients will be terminated by this context done.
-func NewClient(ctx context.Context, clientFactory client.Factory, dialOptions ...grpc.DialOption) networkservice.NetworkServiceClient {
+func NewClient(ctx context.Context, dialTimeout time.Duration, clientFactory client.Factory, dialOptions ...grpc.DialOption) networkservice.NetworkServiceClient {
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/clienturl/client.go:51]: I'm not sure that we should pass this via input parameters. Probably we need to consider have a dynamic `dialTimeout` based on the current load.

Also, I think it is critical to add documentation or create a separate issue with clarification about the problems with GC, that you pointed on last PR meeting.

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/networkservice/chains/nsmgr/server_test.go (hunk 2 above)
Comment (Bolodya1997) [reply pkg/networkservice/chains/nsmgr/server_test.go]: Done.

File: pkg/networkservice/chains/nsmgr/server.go (hunk 1 above)
Comment (Bolodya1997) [reply pkg/networkservice/chains/nsmgr/server.go]: Actually `grpc.WithTimeout` option serves exactly the purpose we need but it is deprecated :(
So OK, now there are `connect.WithDialTimeout` and `connect.WithDialOptions` :)

File: pkg/networkservice/common/clienturl/client.go [hunk 4]
```diff
@@ -77,11 +80,11 @@ func (u *clientURLClient) init() error {
 			return
 		}
 
-		ctx, cancel := context.WithTimeout(u.ctx, dialTimeout)
+		ctx, cancel := context.WithTimeout(u.ctx, u.dialTimeout)
 		defer cancel()
 
 		var cc *grpc.ClientConn
-		cc, u.dialErr = grpc.DialContext(ctx, grpcutils.URLToTarget(clientURL), u.dialOptions...)
+		cc, u.dialErr = grpc.DialContext(ctx, grpcutils.URLToTarget(clientURL), append(u.dialOptions, grpc.WithReturnConnectionError())...)
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/clienturl/client.go:87]: Why do we need this WithReturnConnectionError?

File: pkg/networkservice/common/clienturl/client.go (hunk 3 above)
Comment (denis-tingaikin) [inline pkg/networkservice/common/clienturl/client.go:51]: Why it is not an option?

File: pkg/networkservice/common/connect/server_test.go [hunk 5]
```diff
@@ -326,9 +328,9 @@ func TestConnectServer_RequestParallel(t *testing.T) {
 	wg.Wait()
 	wg.Add(parallelCount)
 
-	require.Equal(t, int32(parallelCount), serverClient.count)
-	require.Equal(t, int32(parallelCount), serverA.count)
-	require.Equal(t, int32(parallelCount), serverNext.count)
+	assert.Equal(t, int32(parallelCount), serverClient.count)
+	assert.Equal(t, int32(parallelCount), serverA.count)
+	assert.Equal(t, int32(parallelCount), serverNext.count)
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/connect/server_test.go:333]: Why we change it here and don't change below?

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/networkservice/common/clienturl/client.go (hunk 4 above)
Comment (denis-tingaikin) [reply pkg/networkservice/common/clienturl/client.go:87]: OK, this is useful https://github.com/grpc/grpc-go/blob/master/dialoptions.go#L293-L301

File: pkg/networkservice/common/clienturl/client.go (hunk 3 above)
Comment (denis-tingaikin) [reply pkg/networkservice/common/clienturl/client.go:51]: Lets point to the issue in the source code

File: pkg/networkservice/common/connect/server_test.go (hunk 5 above)
Comment (Bolodya1997) [reply pkg/networkservice/common/connect/server_test.go:333]: If test doesn't reach `barrier.Done()` (line 335) it breaks, so we can't use here `require` and need `assert`.

File: pkg/networkservice/common/clienturl/client.go (hunk 3 above)
Comment (Bolodya1997) [reply pkg/networkservice/common/clienturl/client.go]: Moved `clienturl` client to `connect` and made it private.

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**